router/batchrouter/asyncdestinationmanager/bing-ads/util.go

```diff
@@ -0,0 +1,444 @@
+	"strings"
+
+	"github.com/google/uuid"
+	"github.com/samber/lo"
+
+	"github.com/rudderlabs/rudder-go-kit/stats"
+	"github.com/rudderlabs/rudder-server/utils/misc"
+)
+
+// Upload related utils
+
+// returns the clientID struct
+func newClientID(jobID int64, hashedEmail string) ClientID {
+	return ClientID{
+		JobID:       jobID,
+		HashedEmail: hashedEmail,
+	}
+}
+
+// returns the string representation of the clientID struct which is of format
+// jobId<<>>hashedEmail
+func (c *ClientID) ToString() string {
+	return fmt.Sprintf("%d%s%s", c.JobID, clientIDSeparator, c.HashedEmail)
+}
+
+/*
+returns the csv file and zip file path, along with the csv writer that
+contains the template of the uploadable file.
+*/
+func createActionFile(audienceId, actionType string) (*ActionFileInfo, error) {
+	localTmpDirName := fmt.Sprintf(`/%s/`, misc.RudderAsyncDestinationLogs)
+	tmpDirPath, err := misc.CreateTMPDIR()
+	if err != nil {
+		return nil, err
+	}
+	path := path.Join(tmpDirPath, localTmpDirName, uuid.NewString())
+	csvFilePath := fmt.Sprintf(`%v.csv`, path)
+	zipFilePath := fmt.Sprintf(`%v.zip`, path)
+	csvFile, err := os.Create(csvFilePath)
+	if err != nil {
+		return nil, err
+	}
+	csvWriter := csv.NewWriter(csvFile)
+	err = csvWriter.WriteAll([][]string{
+		{"Type", "Status", "Id", "Parent Id", "Client Id", "Modified Time", "Name", "Description", "Scope", "Audience", "Action Type", "Sub Type", "Text"},
+		{"Format Version", "", "", "", "", "", "6.0", "", "", "", "", "", ""},
+		{"Customer List", "", audienceId, "", "", "", "", "", "", "", actionType, "", ""},
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &ActionFileInfo{
+		Action:      actionType,
+		ZipFilePath: zipFilePath,
+		CSVFilePath: csvFilePath,
+		CSVWriter:   csvWriter,
+	}, nil
+}
+
+func convertCsvToZip(actionFile *ActionFileInfo) error {
+	if actionFile.EventCount == 0 {
+		os.Remove(actionFile.CSVFilePath)
+		os.Remove(actionFile.ZipFilePath)
+		return nil
+	}
+	zipFile, err := os.Create(actionFile.ZipFilePath)
+	if err != nil {
+		return err
+	}
+	defer zipFile.Close()
+
+	zipWriter := zip.NewWriter(zipFile)
+
+	csvFileInZip, err := zipWriter.Create(filepath.Base(actionFile.CSVFilePath))
+	if err != nil {
+		return err
+	}
+	csvFile, err := os.Open(actionFile.CSVFilePath)
+	if err != nil {
+		return err
+	}
+	if _, err := csvFile.Seek(0, 0); err != nil {
+		return err
+	}
+
+	if _, err = io.Copy(csvFileInZip, csvFile); err != nil {
+		return err
+	}
+
+	// Close the ZIP writer
+	if err = zipWriter.Close(); err != nil {
+		return err
+	}
+	// Remove the csv file after creating the zip file
+	if err = os.Remove(actionFile.CSVFilePath); err != nil {
+		return err
+	}
+	return nil
+}
+
+/*
+Populates the csv file only if it is within the file size limit 100mb and row number limit 4000000
+Otherwise event is appended to the failedJobs and will be retried.
+*/
+func (b *BingAdsBulkUploader) populateZipFile(actionFile *ActionFileInfo, audienceId, line string, data Data) error {
```

Comment:
[minor] method documentation in go should be with `//`, and the comment should start with method name `populateZipFile`
https://tip.golang.org/doc/comment
```suggestion
// populateZipFile only if it is within the file size limit 100mb and row number limit 4000000
// Otherwise event is appended to the failedJobs and will be retried.
func (b *BingAdsBulkUploader) populateZipFile(actionFile *ActionFileInfo, audienceId, line string, data Data) error {
```